mayastor/src/persistent_store.rs

```diff
@@ -0,0 +1,250 @@
+//!
+//! etcd is used as the backing store and is interacted with through the use of
+//! the etcd-client crate. This crate has a dependency on the tokio async
+//! runtime.
+use crate::{
+    core,
+    store::{
+        etcd::Etcd,
+        store_defs::{
+            DeleteWait,
+            GetWait,
+            PutWait,
+            Store,
+            StoreError,
+            StoreKey,
+            StoreValue,
+        },
+    },
+};
+use futures::channel::oneshot;
+use once_cell::sync::OnceCell;
+use serde_json::Value;
+use snafu::ResultExt;
+use std::{future::Future, sync::Mutex, time::Duration};
+
+static DEFAULT_PORT: &str = "2379";
+static STORE_OP_TIMEOUT: Duration = Duration::from_secs(30);
+static PERSISTENT_STORE: OnceCell<Option<Mutex<PersistentStore>>> =
+    OnceCell::new();
+
+/// Persistent store
+pub struct PersistentStore {
+    /// Backing store used for persistence.
+    store: Etcd,
+    /// Endpoint of the backing store.
+    endpoint: String,
+}
+
+impl PersistentStore {
+    /// Initialise the persistent store.
+    /// If the supplied endpoint is 'None', the store is uninitalised and
+    /// unavailable for use.
+    pub async fn init(endpoint: Option<String>) {
+        if endpoint.is_none() {
+            // No endpoint means no persistent store.
+            warn!("Persistent store not initialised");
+            return;
+        }
+
+        assert!(endpoint.is_some());
+
+        // An endpoint has been provided, initialise the persistent store.
+        let endpoint = Self::format_endpoint(&endpoint.unwrap());
+        let store = Self::connect_to_backing_store(&endpoint.clone()).await;
+        PERSISTENT_STORE.get_or_init(|| {
+            Some(Mutex::new(PersistentStore {
+                store,
+                endpoint,
+            }))
+        });
+    }
+
+    /// Adds the default port to the endpoint if one isn't already specified.
+    fn format_endpoint(endpoint: &str) -> String {
+        match endpoint.contains(':') {
+            true => endpoint.to_string(),
+            false => format!("{}:{}", endpoint, DEFAULT_PORT),
+        }
+    }
+
+    /// Connect to etcd as the backing store.
+    /// A connection to the store will be attempted continuously until
+    /// successful. This is necessary as the backing store is essential to the
+    /// operation of Mayastor across restarts.
+    async fn connect_to_backing_store(endpoint: &str) -> Etcd {
+        let mut output_err = true;
+        loop {
+            match Etcd::new(endpoint).await {
+                Ok(store) => {
+                    info!("Connected to etcd on endpoint {}", endpoint);
+                    return store;
+                }
+                Err(_) => {
+                    if output_err {
+                        // Only output the error on first failure to prevent
+                        // flooding the logs.
+                        error!(
+                            "Failed to connect to etcd on endpoint {}",
```

Comment:
Maybe add a "retrying.." to the message?

Reply:
Done.